### PR TITLE
added float32 schubfach algorithm; wip

### PR DESCRIPTION
**Pull request by Araq**

### CODE REVIEW

**Comment by planetis-m** [comment]
There is a float32 implementation of Dragonbox in [fmtlib](https://github.com/fmtlib/fmt/blob/54f22a3eef27c1e29cf0158945ff50bbab232d27/include/fmt/format-inl.h#L828) Did you miss this [comment](https://github.com/nim-lang/Nim/pull/18008#issuecomment-841638502)?

> Alternatively you might want to try out the Schubfach implementation (float32 version, float64 version). Dragonbox is an optimized version of this algorithm.

nvm.

File: lib/system/schubfach.nim [hunk 1]
```diff
@@ -0,0 +1,451 @@
+##  Copyright 2020 Alexander Bolz
```
Comment (timotheecour) [inline lib/system/schubfach.nim:1]: as i was doing in https://github.com/nim-lang/Nim/issues/18008, can you add a readme.md somewhere (maybe `lib/vendor/readme.md` in which we can add future external dependencies under lib/vendor) to show:
* the git repo involved, the files that were copied, and the hash at which the repo was cloned
* the c2nim command used (i'm assuming smthg like `c2nim --cpp --keepBodies -o:lib/system/schubfach.nim $Drachennest/src/schubfach_32.cc`)
* ideally, split in 2 commits, 1 with raw c2nim output (maybe marked `[wip]` so git bisect can learn to skip it), and one with the manual modifications, that way it's more reproducible/transparent and easier to update from upstream if needed

File: lib/system/strmantle.nim [hunk 2]
```diff
@@ -164,6 +164,19 @@ proc nimFloatToStr(f: float): string {.compilerproc.} =
   result = newStringOfCap(8)
   result.addFloat f
 
+when not defined(nimLegacyAddFloat) and not defined(nimscript) and
+    not defined(js) and defined(nimHasDragonBox):
+  import schubfach
+
+proc nimFloat32ToStr(f: float32): string {.compilerproc.} =
```
Comment (timotheecour) [inline lib/system/strmantle.nim:171]: how about adding float32 overloads for `addFloat`, `$`, `writeFloatToBuffer`

File: lib/system/schubfach.nim (hunk 1 above)
Comment (Araq) [reply lib/system/schubfach.nim:1]: Good ideas, will do it maybe later. But I consider remaining bugs to be our own, not related to dragonbox/schubfach, just like we avoided certain TLSF bugs by porting over to Nim.

Comment (timotheecour) [reply lib/system/schubfach.nim:1]: i was referring to not just bugs but performance improvements, not that they can't be ported over again

> Good ideas, will do it maybe later.

=> https://github.com/nim-lang/Nim/commit/282d61bafa4f1b06e1328cf8e517f556fb86d592

**Pull request merged**